packages/vsx-registry/src/common/vsx-registry-api.ts

```diff
@@ -78,4 +88,57 @@ export class VSXRegistryAPI {
         return fetchText(url);
     }
 
+    /**
+     * Get the latest compatible extension version.
+     * - an extension satisfies compatibility if its `engines.vscode` version is supported.
+     * @param id the extension id.
+     *
+     * @returns the data for the latest compatible extension version if available, else `undefined`.
+     */
+    async getLatestCompatibleExtension(id: string): Promise<VSXExtensionRaw | undefined> {
+        const extension = await this.getExtension(id);
+        for (const extensionVersion in extension.allVersions) {
+            if (extensionVersion === 'latest') {
+                continue;
+            }
+            const apiUri = await this.environment.getRegistryApiUri();
+            const data: VSXExtensionRaw = await this.fetchJson(apiUri.resolve(id.replace('.', '/')).toString() + `/${extensionVersion}`);
+            if (data.engines && await this.isEngineValid(data.engines.vscode)) {
+                return data;
+            }
+        }
+    }
+
+    /**
+     * Get the latest compatible version of an extension.
+     * @param versions the `allVersions` property.
+     *
+     * @returns the latest compatible version of an extension if it exists, else `undefined`.
+     */
+    async getLatestCompatibleVersion(id: string, versions: VSXAllVersions[]): Promise<VSXExtensionRaw | undefined> {
+        for (const data of versions) {
+            if (!data.engines) {
+                continue;
+            }
+            if (await this.isEngineValid(data.engines.vscode)) {
+                return this.getExtensionVersion(id, data.version);
+            }
+        }
+        return undefined;
+    }
+
+    /**
+     * Determine if the engine is valid.
+     * @param engine the engine.
+     *
+     * @returns `true` if the engine satisfies the API version.
+     */
+    protected async isEngineValid(engine?: string): Promise<boolean> {
+        if (!engine) {
+            return false;
+        }
+        const apiVersion = await this.variablesServer.getVscodeApiVersion();
```

Comment:
we should fetch it once and memorize in `VSXEnvironment`

Reply:
> we should fetch it once and memorize in `VSXEnvironment`

I believe it is fetched once with the cache present in  [`VSXVariablesServerCached`](https://github.com/eclipse-theia/theia/pull/8191/files/a93170b7a3639858d47840ccebeb290faa3fdcc1#diff-e30b9eb5073fe15c17f1a5bd871e5df8R20-R35), do we need to memorize it as well in `VSXEnvironment`?

Reply:
I think it would be more aligned with as before and `VSXVariablesServerCached` can be removed. Frontend code should rather use `VSXEnvironment` and `VSXVariablesServer` will it be its secret. I would also rename it to `VSXEnvironmentServer`.

Reply:
I updated the code to make use of `VSXEnvironment` (caching it as well).
I believe there are errors at the moment with this new approach when loading plugins (viewed by the CI), so I'll try to investigate further.